src/Interpreters/GinFilter.cpp

```diff
@@ -0,0 +1,224 @@
+#include <string>
+#include <algorithm>
+#include <city.h>
+#include <Columns/ColumnArray.h>
+#include <Columns/ColumnNullable.h>
+#include <Columns/ColumnLowCardinality.h>
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypeNullable.h>
+#include <DataTypes/DataTypeLowCardinality.h>
+#include <Storages/MergeTree/MergeTreeIndexGin.h>
+#include <Storages/MergeTree/MergeTreeIndexFullText.h>
+#include <Disks/DiskLocal.h>
+#include <Storages/MergeTree/GinIndexStore.h>
+#include <Interpreters/GinFilter.h>
+
+using namespace std;
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+}
+GinFilterParameters::GinFilterParameters(size_t ngrams_)
+    : ngrams(ngrams_)
+{
+    if (ngrams > 8)
+        throw Exception("The size of gin filter cannot be greater than 8", ErrorCodes::BAD_ARGUMENTS);
+}
+
+GinFilter::GinFilter(const GinFilterParameters & params)
+    : ngrams(params.ngrams)
+{
+}
+
+void GinFilter::add(const char* data, size_t len, UInt32 rowID, GinIndexStorePtr& store)
+{
+    if (len > MAX_TERM_LENGTH)
+        return;
+
+    string token(data, len);
+    auto it(store->getPostings().find(token));
+
+    if (it != store->getPostings().end())
+    {
+        if (!it->second->contains(rowID))
+            it->second->add(rowID);
+    }
+    else
+    {
+        GinIndexPostingsBuilderPtr builder = std::make_shared<GinIndexPostingsBuilder>();
+        builder->add(rowID);
+
+        store->getPostings()[token] = builder;
+    }
+}
+
+void GinFilter::addRowRangeToGinFilter(UInt32 segmentID, UInt32 rowIDStart, UInt32 rowIDEnd)
+{
+    if (!rowid_range_container.empty())
+    {
+        /// Try to merge the rowID range with the last one in the container
+        if (rowid_range_container.back().segment_id == segmentID &&
+            rowid_range_container.back().range_end+1 == rowIDStart)
+        {
+            rowid_range_container.back().range_end = rowIDEnd;
+            return;
+        }
+    }
+    rowid_range_container.push_back({segmentID, rowIDStart, rowIDEnd});
+}
+
+void GinFilter::clear()
+{
+    rowid_range_container.clear();
```

Comment:
Should that not clear the other members as well?

Reply:
Fixed.

Reply:
Fixed.